packages/router/src/adapters/contract/index.ts

```diff
@@ -1,47 +1,85 @@
 import { BigNumber, providers } from "ethers/lib/ethers";
 
-import { prepare, fulfill, cancel, removeLiquidity, getRouterBalance, sanitationCheck } from "./contract";
+import { startContractListeners } from "../../lib/helpers";
+
+import {
+  prepareTransactionManager,
+  prepareRouterContract,
+  fulfillTransactionManager,
+  fulfillRouterContract,
+  cancelRouterContract,
+  cancelTransactionManager,
+  removeLiquidityTransactionManager,
+  getRouterBalance,
+} from "./contract";
 
 export type ContractWriter = {
-  prepare: (
+  prepareTransactionManager: (
     chainId: number,
     prepareParams: PrepareParams,
     requestContext: RequestContext,
   ) => Promise<providers.TransactionReceipt>;
-  fulfill: (
+  prepareRouterContract: (
+    chainId: number,
+    prepareParams: PrepareParams,
+    routerContractAddress: string,
+    signature: string,
+    routerRelayerFeeAsset: string,
+    routerRelayerFee: string,
+    useRelayer: boolean,
+    requestContext: RequestContext,
+  ) => Promise<providers.TransactionReceipt>;
+  fulfillTransactionManager: (
+    chainId: number,
+    fulfillParams: FulfillParams,
+    requestContext: RequestContext,
+  ) => Promise<providers.TransactionReceipt>;
+  fulfillRouterContract: (
     chainId: number,
     fulfillParams: FulfillParams,
+    routerContractAddress: string,
+    signature: string,
+    routerRelayerFeeAsset: string,
+    routerRelayerFee: string,
+    useRelayer: boolean,
     requestContext: RequestContext,
   ) => Promise<providers.TransactionReceipt>;
-  cancel: (
+  cancelTransactionManager: (
     chainId: number,
     cancelParams: CancelParams,
     requestContext: RequestContext,
   ) => Promise<providers.TransactionReceipt>;
-  removeLiquidity: (
+  cancelRouterContract: (
+    chainId: number,
+    cancelParams: CancelParams,
+    routerContractAddress: string,
+    signature: string,
+    routerRelayerFeeAsset: string,
+    routerRelayerFee: string,
+    useRelayer: boolean,
+    requestContext: RequestContext,
+  ) => Promise<providers.TransactionReceipt>;
+  removeLiquidityTransactionManager: (
     chainId: number,
     amount: string,
     assetId: string,
     recipientAddress: string | undefined,
     requestContext: RequestContext,
   ) => Promise<providers.TransactionReceipt>;
   getRouterBalance: (chainId: number, router: string, assetId: string) => Promise<BigNumber>;
-  sanitationCheck: (
-    chainId: number,
-    transactionData: TransactionData,
-    functionCall: "prepare" | "fulfill" | "cancel",
-    _requestContext?: RequestContext<string>,
-  ) => Promise<void>;
 };
 
 export const contractWriter = (): ContractWriter => {
+  startContractListeners();
   return {
-    prepare,
-    fulfill,
-    cancel,
-    removeLiquidity,
+    prepareTransactionManager,
+    prepareRouterContract,
+    fulfillTransactionManager,
+    fulfillRouterContract,
+    cancelRouterContract,
+    cancelTransactionManager,
+    removeLiquidityTransactionManager,
```

Comment:
can't you do an internal check to see if you should use the `prepareTransactionManager` over the `prepareRouterContract` for each of these functions so it would just show up as `prepare` and handle the checks internally?

Reply:
i think this may be in place in the `operations`?

Reply:
Yes this is definitely in place but it's not that easy because you need to send relayer fees in some cases so there's more business logic that should be in operations.